aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -22,12 +22,41 @@ using namespace at::sparse;
 
 namespace {
 
-  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
-    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
+  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input, bool promotes_integer_to_float = false) {
+    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true) && promotes_integer_to_float) {
       return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
     }
     return at::empty_like(input);
   }
+
+  // --------------------------------------------------------------------
+  // helper function for sanity check of input and output tensors
+  // this helper function is for unary ops
+  // --------------------------------------------------------------------
+  static inline void sparse_unary_op_check_sanity(SparseTensor& result, const SparseTensor& self,
+      bool promotes_integer_to_float = false, bool is_linear_map = false) {
+    TORCH_CHECK(result.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.get_device() == result.get_device(),
+        "Both tensors should be on the same device");
+
+    if (promotes_integer_to_float){
+      TORCH_CHECK(
+        !c10::isIntegralType(result.scalar_type(), /*includeBool=*/true),
+        "result type cannot be Integral, got:",
+        result.scalar_type());
+    }
+
+    if (is_same_tensor(result, self) && is_linear_map) {
```

Comment:
Shouldn't this check be `&& !is_linear_map` -- because it would be safe to perform the operation if it was a linear map, right?

Reply:
@pearu @mruberry - Made the changes as suggested